webrtc/RTCRtpTransceiver-setCodecPreferences.html

```diff
@@ -0,0 +1,136 @@
+    const capabilities1 = RTCRtpSender.getCapabilities('audio');
+    const capabilities2 = RTCRtpReceiver.getCapabilities('audio');
+    transceiver.setCodecPreferences([...capabilities1.codecs, ... capabilities2.codecs]);
+
+  }, `setCodecPreferences() with both sender receiver codecs combined should succeed`);
+
+  test(() => {
+    const pc = new RTCPeerConnection();
+    const transceiver = pc.addTransceiver('audio');
+    transceiver.setCodecPreferences([]);
+
+  }, `setCodecPreferences([]) should succeed`);
+
+  test(() => {
+    const pc = new RTCPeerConnection();
+    const transceiver = pc.addTransceiver('audio');
+    const capabilities = RTCRtpSender.getCapabilities('audio');
+    const { codecs } = capabilities;
+
+    if(codecs.length >= 2) {
+      const tmp = codecs[0];
+      codecs[0] = codecs[1];
+      codecs[1] = tmp;
+    }
+
+    transceiver.setCodecPreferences(codecs);
+
+  }, `setCodecPreferences() with reordered codecs should succeed`);
+
+  test(() => {
+    const pc = new RTCPeerConnection();
+    const transceiver = pc.addTransceiver('audio');
+    const capabilities = RTCRtpSender.getCapabilities('video');
+    assert_throws(() => transceiver.setCodecPreferences(capabilities.codecs));
+
+  }, `setCodecPreferences() on audio transceiver with codecs returned from getCapabilities('video') should throw InvalidAccessError`);
+
+  test(() => {
+    const pc = new RTCPeerConnection();
+    const transceiver = pc.addTransceiver('audio');
+    const codecs = [{
+      mimeType: 'audio/piepiper',
+      clockRate: 2000,
+      channels: 2,
+      sdpFmtpLine: 'a=fmtp:98 0-15'
+    }];
+
+    assert_throws(() => transceiver.setCodecPreferences(codecs));
+
+  }, `setCodecPreferences() with user defined codec should throw InvalidAccessError`);
+
+  test(() => {
+    const pc = new RTCPeerConnection();
+    const transceiver = pc.addTransceiver('audio');
+    const capabilities = RTCRtpSender.getCapabilities('audio');
+    const codecs = [
+      ...capabilities.codecs,
+      {
+        mimeType: 'audio/piepiper',
+        clockRate: 2000,
+        channels: 2,
+        sdpFmtpLine: 'a=fmtp:98 0-15'
+      }];
+
+    assert_throws(() => transceiver.setCodecPreferences(codecs));
+
+  }, `setCodecPreferences() with user defined codec together with codecs returned from getCapabilities() should throw InvalidAccessError`);
+
+  test(() => {
+    const pc = new RTCPeerConnection();
+    const transceiver = pc.addTransceiver('audio');
+    const capabilities = RTCRtpSender.getCapabilities('audio');
+
+    const { codecs } = capabilities;
+    assert_greater_than(codecs.length, 0,
+      'Expect at least one codec available');
+
+    const [ codec ] = codecs;
+    const { channels=2 } = codec;
+    codec.channels = channels+1;
+
+    assert_throws(() => transceiver.setCodecPreferences(codecs));
+
+  }, `setCodecPreferences() with modified codecs returned from getCapabilities() should throw InvalidAccessError`);
```

Comment:
I think it would always get read, since dictionary is passed by value in WebIDL. It is pretty tough to attempt modifying the fields because they may be undefined, and we can't just simply assign an arbitrary constant value because we don't know if it would be the same as the old value.